File: smallest-number-in-infinite-set/smallest-number-in-infinite-set.java
class SmallestInfiniteSet {
    int i = 1;
    HashSet<Integer> isPresent = new HashSet<>();
    PriorityQueue<Integer> addedBack = new PriorityQueue<>();
    public SmallestInfiniteSet() {
        
    }
    
    public int popSmallest() {
        int answer;
        if (!addedBack.isEmpty()) {
            answer = addedBack.poll();
            isPresent.remove(answer);
        }
        else {
            answer = i++;
        }
        return answer;
    }
    
    public void addBack(int num) {
        if (i <= num || isPresent.contains(num)) {
            return;
        }
        addedBack.add(num);
        isPresent.add(num);
    }
}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet obj = new SmallestInfiniteSet();
 * int param_1 = obj.popSmallest();
 * obj.addBack(num);
 */

File: 2413-smallest-number-in-infinite-set/smallest-number-in-infinite-set.py
class SmallestInfiniteSet:

    def __init__(self):
        self.heap = []
        self.pointer = 1
        self.unique_heap = set()
    def popSmallest(self) -> int:
        if not self.heap:
            ret = self.pointer
            self.pointer += 1
            return ret
        else:
            val = heapq.heappop(self.heap)
            self.unique_heap.remove(val)
            return val
    def addBack(self, num: int) -> None:
        if self.pointer > num and num not in self.unique_heap:
            heapq.heappush(self.heap, num)
            self.unique_heap.add(num)


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)